Description: do not set ArgsEscaped (on Linux)
Author: Tianon Gravi <tianon@debian.org>
Forwarded: https://github.com/moby/buildkit/pull/4723

diff --git a/frontend/dockerfile/dockerfile2llb/convert.go b/frontend/dockerfile/dockerfile2llb/convert.go
index 403530a74..ccd6c105a 100644
--- a/frontend/dockerfile/dockerfile2llb/convert.go
+++ b/frontend/dockerfile/dockerfile2llb/convert.go
@@ -1399,7 +1399,7 @@ func dispatchCmd(d *dispatchState, c *instructions.CmdCommand) error {
 		args = withShell(d.image, args)
 	}
 	d.image.Config.Cmd = args
-	d.image.Config.ArgsEscaped = true //nolint:staticcheck // ignore SA1019: field is deprecated in OCI Image spec, but used for backward-compatibility with Docker image spec.
+	// TODO set ArgsEscaped appropriately *ONLY FOR WINDOWS* (it should never ever be set for Linux)
 	d.cmdSet = true
 	return commitToHistory(&d.image, fmt.Sprintf("CMD %q", args), false, nil, d.epoch)
 }
