Description: disable recursive cloning of submodules by default
Forwarded: no (upstream's default makes sense for them)

diff --git a/frontend/dockerfile/dockerfile2llb/convert_copy.go b/frontend/dockerfile/dockerfile2llb/convert_copy.go
index 37c3d4e7e..a55a3a086 100644
--- a/frontend/dockerfile/dockerfile2llb/convert_copy.go
+++ b/frontend/dockerfile/dockerfile2llb/convert_copy.go
@@ -164,7 +164,7 @@ func dispatchCopy(d *dispatchState, cfg copyConfig) error {
 			if gitRef.SubDir != "" {
 				gitOptions = append(gitOptions, llb.GitSubDir(gitRef.SubDir))
 			}
-			if gitRef.Submodules != nil && !*gitRef.Submodules {
+			if gitRef.Submodules == nil || !*gitRef.Submodules {
 				gitOptions = append(gitOptions, llb.GitSkipSubmodules())
 			}
 
diff --git a/frontend/dockerui/context.go b/frontend/dockerui/context.go
index 3bd5094f6..27c34f454 100644
--- a/frontend/dockerui/context.go
+++ b/frontend/dockerui/context.go
@@ -169,7 +169,7 @@ func DetectGitContext(ref string, keepGit *bool, opts ...llb.GitOption) (*llb.St
 	if g.Checksum != "" {
 		gitOpts = append(gitOpts, llb.GitChecksum(g.Checksum))
 	}
-	if g.Submodules != nil && !*g.Submodules {
+	if g.Submodules == nil || !*g.Submodules {
 		gitOpts = append(gitOpts, llb.GitSkipSubmodules())
 	}
 	if g.MTime != "" {
