Description: disable recursive cloning of submodules by default
Forwarded: no (upstream's default makes sense for them)

diff --git a/frontend/dockerfile/dockerfile2llb/convert.go b/frontend/dockerfile/dockerfile2llb/convert.go
index 665bdb399..271ef0852 100644
--- a/frontend/dockerfile/dockerfile2llb/convert.go
+++ b/frontend/dockerfile/dockerfile2llb/convert.go
@@ -1545,7 +1545,7 @@ func dispatchCopy(d *dispatchState, cfg copyConfig) error {
 			if gitRef.SubDir != "" {
 				gitOptions = append(gitOptions, llb.GitSubDir(gitRef.SubDir))
 			}
-			if gitRef.Submodules != nil && !*gitRef.Submodules {
+			if gitRef.Submodules == nil || !*gitRef.Submodules {
 				gitOptions = append(gitOptions, llb.GitSkipSubmodules())
 			}
 
diff --git a/frontend/dockerui/context.go b/frontend/dockerui/context.go
index 88d68569b..19a3a0b88 100644
--- a/frontend/dockerui/context.go
+++ b/frontend/dockerui/context.go
@@ -164,7 +164,7 @@ func DetectGitContext(ref string, keepGit *bool) (*llb.State, bool, error) {
 	if g.Checksum != "" {
 		gitOpts = append(gitOpts, llb.GitChecksum(g.Checksum))
 	}
-	if g.Submodules != nil && !*g.Submodules {
+	if g.Submodules == nil || !*g.Submodules {
 		gitOpts = append(gitOpts, llb.GitSkipSubmodules())
 	}
 
